esp: Added basic implemetation of message passing structure

jhonnyam opened the pull request:
Implementation of a linux list.h and callback based implementation for flexible message passing.

- [x] provide basics for implementation
- [x] Add macros for easy initialization
- [x] Add default callbacks for usage with queue
- [x] Add handling of push result in case of empty queues etc.
- [x] tested

Commit:
esp:feature added channel component

nopeslide commented on include/message.h:
```diff
@@ -0,0 +1,29 @@
+#ifndef __MESSAGE__
+#define __MESAAGE
```
```suggestion
#define __MESSAGE__
```

nopeslide commented on include/message.h:
```diff
@@ -0,0 +1,29 @@
+    const char *name;
+    struct list_head same;
+    struct list_head unique;
+    void *state;
```
name is confusing
```suggestion
    void *ctx;
```

nopeslide commented on include/message.h:
```diff
@@ -0,0 +1,29 @@
+} channel_group;
+
+
+void init_message();
```
no message anywhere, did you mean channel?
```suggestion
void channel_init();
```

nopeslide commented on include/message.h:
```diff
@@ -0,0 +1,29 @@
+
+void init_message();
+
+void register_channel(channel_group *group);
```
```suggestion
void channel_register(channel_group *group);
```

nopeslide commented on include/message.h:
```diff
@@ -0,0 +1,29 @@
+void init_message();
+
+void register_channel(channel_group *group);
+void unregister_channel(channel_group *group);
```
```suggestion
void channel_unregister(channel_group *group);
```

nopeslide commented on include/message.h:
```diff
@@ -0,0 +1,29 @@
+void register_channel(channel_group *group);
+void unregister_channel(channel_group *group);
+
+void publish(channel_group *channel, void *data, unsigned int size);
```
```suggestion
void channel_publish(channel_group *channel, void *data, unsigned int size);
```

nopeslide commented on src/message.c:
```diff
@@ -0,0 +1,88 @@
+    INIT_LIST_HEAD(&unique_head.unique);
+    unique_head.push = NULL;
+    unique_head.state = NULL;
+    unique_head.data_size = 0;
```
i would replace this with a macro

ikstream replied:
Please use functions, where ever possible instead of Macros

nopeslide replied:
we provide a function utilizing the macro, but a macro is preferable at this point, since it can be used in non-code segments as static initialization.

nopeslide commented on include/message.h:
```diff
@@ -0,0 +1,29 @@
+    struct list_head unique;
+    void *state;
+    unsigned int data_size;
+    push_result (*push) (struct group *, void *, unsigned int);
```
```suggestion
    callback_status (*callback) (struct group *, void *, unsigned int);
```

nopeslide commented on include/message.h:
```diff
@@ -0,0 +1,29 @@
+
+typedef enum {
+    PUSH_OK, PUSH_ERROR, PUSH_BLOCK
+} push_result;
```
```suggestion
typedef enum {
    CALLBACK_OK, CALLBACK_ERROR, CALLBACK_BLOCK
} callback_status;
```

nopeslide commented on include/message.h:
```diff
@@ -0,0 +1,29 @@
+    struct list_head same;
+    struct list_head unique;
+    void *state;
+    unsigned int data_size;
```
lets remove `data_size` and assume we only push single event each time

nopeslide replied:
why is this even necessary?

nopeslide replied:
shouldn't this be actually handled by the callback?

nopeslide commented:
`message.{c,h}` have no actual message object, maybe rename it to `channel` or `event`? further, `channel_group` to simple `channel` or `event`?

nopeslide commented on include/channel.h:
```diff
@@ -3,20 +3,25 @@
+typedef int TickType_t;
+#define pdPass 1;
+void xQueueGenericSend(void *ctx, const void * const data, TickType_t timeout, const BaseType_t flags);
+
```
will be changed, is only there for the compile dry-run

nopeslide commented:
needs to be tested and documentation needs to be updated

ikstream commented on include/channel.h:
```diff
@@ -0,0 +1,203 @@
+    void *ctx;
+    BaseType_t flags;
+    Channel_callback callback;
+} Channel;
```
Why is channel uppercase here? Shouldn't it be lowercase as the rest of typedefs?
```suggestion
} channel;
```

nopeslide replied:
this depends on our coding guidelines, should complex types start with lower or uppercase?

ikstream commented on include/list.h:
```diff
@@ -0,0 +1,521 @@
+ * This is only for internal list manipulation where we know
+ * the prev/next entries already!
+ */
+static inline void __list_add(struct list_head *new,
```
In the whole file you have the return type in the same line as the function name
```suggestion
static
void
inline
__list_add(struct list_head *new,
```

ikstream commented on include/list.h:
```diff
@@ -0,0 +1,521 @@
+ * Insert a new entry after the specified head.
+ * This is good for implementing stacks.
+ */
+static inline void list_add(struct list_head *new, struct list_head *head)
```
s.a

ikstream commented on include/list.h:
```diff
@@ -0,0 +1,521 @@
+ * Insert a new entry before the specified head.
+ * This is useful for implementing queues.
+ */
+static inline void list_add_tail(struct list_head *new, struct list_head *head)
```
s.a.

ikstream commented on include/list.h:
```diff
@@ -0,0 +1,521 @@
+ * This is only for internal list manipulation where we know
+ * the prev/next entries already!
+ */
+static inline void __list_del(struct list_head * prev, struct list_head * next)
```
s.a.

ikstream commented on include/list.h:
```diff
@@ -0,0 +1,521 @@
+ * Note: list_empty on entry does not return true after this, the entry is
+ * in an undefined state.
+ */
+static inline void list_del(struct list_head *entry)
```
s.a.

ikstream commented on include/list.h:
```diff
@@ -0,0 +1,521 @@
+ * list_del_init - deletes entry from list and reinitialize it.
+ * @entry: the element to delete from the list.
+ */
+static inline void list_del_init(struct list_head *entry)
```
s.a.

ikstream commented on include/list.h:
```diff
@@ -0,0 +1,521 @@
+ * @list: the entry to move
+ * @head: the head that will precede our entry
+ */
+static inline void list_move(struct list_head *list, struct list_head *head)
```
s.a.

ikstream commented on include/list.h:
```diff
@@ -0,0 +1,521 @@
+ * @list: the entry to move
+ * @head: the head that will follow our entry
+ */
+static inline void list_move_tail(struct list_head *list,
```
s.a.

ikstream commented on include/list.h:
```diff
@@ -0,0 +1,521 @@
+ * list_empty - tests whether a list is empty
+ * @head: the list to test.
+ */
+static inline int list_empty(const struct list_head *head)
```
s.a.

ikstream commented on include/list.h:
```diff
@@ -0,0 +1,521 @@
+	return head->next == head;
+}
+
+static inline void __list_splice(struct list_head *list,
```
s.a.

ikstream commented on include/list.h:
```diff
@@ -0,0 +1,521 @@
+ * @list: the new list to add.
+ * @head: the place to add it in the first list.
+ */
+static inline void list_splice(struct list_head *list, struct list_head *head)
```
s.a.

ikstream commented on include/list.h:
```diff
@@ -0,0 +1,521 @@
+ *
+ * The list at @list is reinitialised
+ */
+static inline void list_splice_init(struct list_head *list,
```
s.a.

ikstream commented on include/list.h:
```diff
@@ -0,0 +1,521 @@
+ * @type:	the type of the struct this is embedded in.
+ * @member:	the name of the list_struct within the struct.
+ */
+#define list_entry(ptr, type, member) \
```
I get why some of these need to be macros, but we should aim to use functions where ever possible. Could any of these be a function?

ikstream commented:
There are more changes needed, but they are mostly cosmetic.

ikstream commented on include/list.h:
```diff
@@ -0,0 +1,521 @@
+	entry->prev = LIST_POISON2;
+}
+
+
```
A lot of new spaces

ikstream commented on include/list.h:
```diff
@@ -0,0 +1,521 @@
+ */
+static inline int list_empty(const struct list_head *head)
+{
+	return head->next == head;
```
beautiful, but could be deceiving

nopeslide commented:
rebased on the new idf structure #50 and moved the `list.h` in its own component titled `linux` to make it clear it's an untouched third_party library

nopeslide committed:
channel:feature refactored, added unit test & debug functions

jhonnyam commented on components/channel/include/channel.h:
```diff
@@ -0,0 +1,230 @@
+ * and calls the callback function on it. May return with error_code before all elements were notified, allowing for custom error handling
+ * Example for skipping failing callbacks:
+ *  Channel ch;
+ *  Channel *pos;
```
Channel *pos isn't needed anymore since it's part of the Broadcast handle

jhonnyam commented on components/channel/test/channel_test.c:
```diff
@@ -0,0 +1,398 @@
+  TEST_ASSERT_LIST_LENGTH(&ch2.same, 3);
+  TEST_ASSERT_LIST_LENGTH(&ch3.same, 2);
+  TEST_ASSERT_LIST_LENGTH(&ch4.same, 2);
+  TEST_ASSERT_LIST_EMPTY (&ch5.same);
```
These tests are duplicate of the register test case.

nopeslide replied:
I left them there to show the state before unregistering

jhonnyam commented on components/channel/test/channel_test.c:
```diff
@@ -0,0 +1,398 @@
+  TEST_ASSERT_LIST_LENGTH(&ch4.same, 2);
+  TEST_ASSERT_LIST_EMPTY (&ch5.same);
+
+  channel_unregister(&ch3);
```
This covers only the most complex path. Maybe add tests for unregister single unique and same channels

jhonnyam reviewed (commented):
Looks good. Some minor details to change.

nopeslide committed:
channel:refactored added Kconfig, changed interface, refactored internals

ikstream commented on components/channel/include/channel_debug.h:
```diff
@@ -0,0 +1,133 @@
+channel_debug_print_dot
+(const char *prefix, const Channel *ch)
+{
+    if (strcmp("",ch->identifier)==0) {
```
add spaces around `==`
```suggestion
    if (strcmp("",ch->identifier) == 0) {
```

ikstream commented on components/channel/src/channel.c:
```diff
@@ -0,0 +1,143 @@
+(Channel *ch) 
+{
+    int isUnique = ! list_empty(&ch->unique);
+    int isSame   = ! list_empty(&ch->same);
```
why are some variables camel case now?

```suggestion
    int is_unique = ! list_empty(&ch->unique);
    int is_same   = ! list_empty(&ch->same);
```

ikstream commented on components/channel/src/channel.c:
```diff
@@ -0,0 +1,143 @@
+    //channel is in unique list and there are others in same list, so replace channel
+    if (isUnique && isSame) {
+        Channel *nextSame = list_entry(ch->same.next, struct channel, same);
+        Channel *nextUniq = list_entry(ch->unique.next, struct channel, unique);
```
```suggestion
        Channel *next_same = list_entry(ch->same.next, struct channel, same);
        Channel *next_uniq = list_entry(ch->unique.next, struct channel, unique);
```

ikstream commented on components/channel/include/channel.h:
```diff
@@ -0,0 +1,211 @@
+ */
+static
+bool
+inline __attribute__((always_inline))
```
as far as i understand \_\_attribute\_\_((always_inline)) prepending inline is not needed
```suggestion
__attribute__((always_inline))
```

ikstream commented on components/channel/src/channel.c:
```diff
@@ -0,0 +1,143 @@
+        return;
+    }
+    //channel is in unique list and there are others in same list, so replace channel
+    if (isUnique && isSame) {
```
```suggestion
    if (!is_unique && !is_same) {
        return;
    }
    //channel is not in the unique chain so just delete it
    if (!is_unique && is_same) {
        list_del(&ch->same);
        INIT_LIST_HEAD(&ch->same);
        return;
    }
    
    //channel is in the unique list, but single element so also just delete it
    if (is_unique && !is_same) {
        list_del(&ch->unique);
        INIT_LIST_HEAD(&ch->unique);
        return;
    }

    //channel is in unique list and there are others in same list, so replace channel
    if (is_unique && is_same) {
```

ikstream commented on components/channel/src/channel.c:
```diff
@@ -0,0 +1,143 @@
+        INIT_LIST_HEAD(&ch->same);
+        list_del(&ch->unique);
+        INIT_LIST_HEAD(&ch->unique);
+        list_add(&nextSame->unique, &nextUniq->unique);
```
```suggestion
        list_add(&next_same->unique, &next_uniq->unique);
```

ikstream reviewed (changes requested):
some minor changes, but so far lgtm.

haven't checked the test file so far

nopeslide committed:
channel:refatored renamed channel_{in,out} to {consumer,producer}

ikstream commented on components/channel/include/channel.h:
```diff
@@ -57,8 +57,10 @@ struct channel {
 typedef struct channel Channel;
-typedef struct channel Channel_in;
-typedef struct channel Channel_out;
+typedef struct channel Channel_in __attribute__((deprecated));
+typedef struct channel Channel_out __attribute__((deprecated));
```
Why use deprecated here? Nothing containing the channel structure has been merged so far. Why not remove it?

ikstream commented on components/channel/include/channel.h:
```diff
@@ -85,14 +87,43 @@ channel_init
+ */
+static
+void
+inline __attribute__((always_inline)) __attribute__((deprecated))
```
why is producer deprecated?
```suggestion
__attribute__((always_inline))
```

Commit:
esp32:feature Added testcase with queue

Commit:
esp32:feature Cleanup queue and task after test

Commit:
esp32:fix Pin test task to core0

Fixes crashes of rerunning tests

Commit:
esp32:fix Remove pin to core. Add task delay instead of busy wait

nopeslide committed:
documentation:fix old function documentation

nopeslide committed:
channel:feature timeout per broadcast not callback

- broadcast accumulates elapsed timeslices in new 'elapsed' counter
- if broadcast takes more time than specified it errors with pdFAIL
- added helper to determine if timeout happend

Commit:
esp32:fix Fixed Testcases for timout and a bug in channel_broadcast_timout

Commit:
esp32:test Added test for channel component with two channels/tasks in a loop

nopeslide commented on components/channel/include/channel.h:
```diff
@@ -0,0 +1,263 @@
+
+typedef struct channel Channel;
+typedef struct channel Channel_in __attribute__((deprecated));
+typedef struct channel Channel_out __attribute__((deprecated));
```
```suggestion
```

Commit:
esp32:fix Removed depricated init_channel_input functions

Commit:
esp32:fix Changed shape and label to differenciate between producer and consumer

ikstream commented on components/channel/include/channel.h:
```diff
@@ -0,0 +1,232 @@
+ */
+static
+bool
+inline __attribute__((always_inline))
```
```suggestion
__attribute__((always_inline))
```

nopeslide replied:
> GCC does not inline any functions when not optimizing unless you specify the ‘always_inline’ attribute for the function, like this:
>
>/* Prototype.  */
> `inline void foo (const char) __attribute__((always_inline));`

see https://gcc.gnu.org/onlinedocs/gcc/Inline.html

ikstream replied:
> > ** 1) As I understood the description of always_inline, the inline
> keyword in the above example is obsolete, right?
> [According to my tests this seems to be the case, but want to be sure :-)]

> If you use the always_inline attribute, then, yes, you do not need to
> also use the inline keyword.  The inline keyword is not in general
> obsolete.

see: https://gcc.gnu.org/legacy-ml/gcc-help/2007-01/msg00051.html

nopeslide replied:
You are right, it seems unnecessary, but will throw warnings:
```
components/channel/include/channel.h:97:1: warning: always_inline function might not be inlinable [-Wattributes]
 channel_init_consumer
 ^~~~~~~~~~~~~~~~~~~~~
esp32-core/components/channel/include/channel.h:97:1: warning: 'channel_init_consumer' defined but not used [-Wunused-function]
```
we could remove the attribute, since we always build with optimizations and then no combination yields a symbol

ikstream commented on components/channel/include/channel.h:
```diff
@@ -0,0 +1,232 @@
+ */
+static
+void
+inline __attribute__((always_inline))
```
```suggestion
__attribute__((always_inline))
```

ikstream commented on components/channel/include/channel.h:
```diff
@@ -0,0 +1,232 @@
+ */
+static
+void
+inline __attribute__((always_inline))
```
```suggestion
__attribute__((always_inline))
```

ikstream commented on components/channel/include/channel.h:
```diff
@@ -0,0 +1,232 @@
+ */
+static
+void
+inline __attribute__((always_inline))
```
```suggestion
__attribute__((always_inline))
```

ikstream commented on components/channel/include/channel.h:
```diff
@@ -0,0 +1,232 @@
+ */
+static
+void 
+inline __attribute__((always_inline))
```
```suggestion
__attribute__((always_inline))
```

ikstream commented on components/channel/include/channel.h:
```diff
@@ -0,0 +1,232 @@
+ */
+static
+void
+inline __attribute__((always_inline))
```
```suggestion
__attribute__((always_inline))
```

ikstream reviewed (changes requested):
looks good to me so far, afar from the points. Not tested yet

nopeslide commented:
@jhonnyam if the qemu simulation fails, but the device test seems to be ok, deactivate the corresponding tests in the `lifesensor_ci/sdkconfig.defaults`

jhonnyam commented:
@nopeslide That feels not right, especially since it's not allways the same test, that fails/crashes

nopeslide commented:
@jhonnyam this was meant for unsolvable tests, if you think you can get it to work pls do so :D

Commit:
esp32:fix Changed testcases to use ints instead of floats as test values

QEMUs support of floats is somewhat broken and leeds to unrialiable test turnouts

Commit:
esp32:refactor Removed allways_inline atrribute since it's not needed

See discussion in #47
https://github.com/ShilaTu/esp32-core/pull/47#discussion_r417804028

Commit:
debug
